[improve][broker] Remove ForkJoin common pool in caffeine cache

zhanghaou opened the pull request:
### Motivation
Same to https://github.com/apache/pulsar/pull/24383

Currently, the usage of Caffeine in pulsar code is inconsistent. Some asynchronous builds use the default ForkJoin common pool, while others specify a custom thread pool, and there are also synchronous builds.

The ForkJoin pool has a specific bug that can prevent tasks from being executed after being submitted to the thread pool. This issue has been fixed in Java 21.0.8, but it still exists in Java 17.
https://bugs.openjdk.org/browse/JDK-8330017
https://bugs.openjdk.org/browse/JDK-8351933

If the ForkJoin common pool gets blocked, it can prevent all dependent caches from fetching data.


### Modifications

1. When executor in pulsarService is available, use pulsarService executor
2. Add new executor in AbstractMetadataStore.java, which use 4 corePoolSize(hard code, need set to be configured?)
3. Use MoreExecutors.directExecutor() for other caffeines.

**The choice of thread pool may not be the most appropriate. Please provide suggestions for improvement, thanks.**

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

zhanghaou commented:
> why is this a problem if there are different ways to use Caffeine?

It’s not necessarily a problem, but I’d like to highlight that certain parts of the code explicitly specify a thread pool, whereas other parts continue to use the default thread pool.

> I understand that it's severe when that happens. How often have you observed this bug? Does it require that that JVM has been running for very long or is it just random when there's heavy load?

It occurs very rarely, only after running for a long time, and it doesn’t seem to be related to the load.

lhotari commented:
> It’s not necessarily a problem, but I’d like to highlight that certain parts of the code explicitly specify a thread pool, whereas other parts continue to use the default thread pool.

For blocking IO, ForkJoinPool is a reasonable solution to avoid blocking Netty IO threads. VirtualThreads in Java 21 could be a better option eventually. Instead of removing ForkJoinPool usage blindly, there should be more analysis of each case so that we don't cause performance regressions. In Pulsar, the execution model isn't clean and more effort should be put in designing how thread pools and task execution should behave in general. This problem has existed for a long time so it's waiting for contributors. :)

nodece commented on pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java:
```diff
@@ -100,9 +101,12 @@ protected AbstractMetadataStore(String metadataStoreName, OpenTelemetry openTele
         this.executor = new ScheduledThreadPoolExecutor(1,
                 new DefaultThreadFactory(
                         StringUtils.isNotBlank(metadataStoreName) ? metadataStoreName : getClass().getSimpleName()));
+        this.cacheExecutor = new ScheduledThreadPoolExecutor(4,
```
Do you consider using CPU cores?

zhanghaou commented:
> For blocking IO, ForkJoinPool is a reasonable solution to avoid blocking Netty IO threads or the Metadata store thread. VirtualThreads in Java 21 could be a better option eventually. 

I agree that VirtualThreads is better. However, given the current situation, ForkJoin common pool is not very suitable.

> Instead of removing ForkJoinPool usage blindly, there should be more analysis of each case so that we don't cause performance regressions. 

https://github.com/apache/pulsar/pull/22589  this pr change executor from ForkJoinPool to pulsar service executor. It seems there’s no analysis regarding performance regression. 

> In Pulsar, the execution model isn't clean and more effort should be put in designing how thread pools and task execution should behave in general. This problem has existed for a long time so it's waiting for contributors. :)

Look forward to it.

lhotari reviewed (changes requested):
I don't support making this change, since it's blindly making changes across the code base. For example in JwksCache, it's necessary to have an actual executor to run tasks in background. The `refreshAfterWrite` will handle that. This PR breaks existing behavior.

zhanghaou commented:
> I don't support making this change, since it's blindly making changes across the code base. For example in JwksCache, it's necessary to have an actual executor to run tasks in background. The `refreshAfterWrite` will handle that. This PR breaks existing behavior.

Added a new executor to address it. Please take a review again. Thanks.

zhanghaou commented:
> I saw this comment just now, it should be fine to wait for the JDK patch, so I dismissed my approval.

Waiting for the JDK upgrade and using the latest version can indeed solve this problem, but wouldn't it be better if we could solve this issue without relying on a JDK upgrade? 

If the user doesn't know that they need to upgrade the JDK to a specified version, there will always be a risk of blocking. That's really scary!

> Using a dedicated executor is reasonable. 

That's the purpose of this PR.

> But the `ThreadPoolExecutor` based executor is not better than ForkJoinPool, which uses work-stealing and is leveraged by the implementation of virtual threads. A simple difference is that if a task takes too long to complete, the queued tasks in the same thread don't have a chance to execute.
> 
> `ForkJoinPool` is a better choice for independent tasks like those tasks to avoid blocking Netty IO threads or the Metadata store thread in Pulsar.

Use `ThreadPoolExecutor` based executor is refer to the existing code https://github.com/apache/pulsar/pull/22589 

@BewareMyPower

BewareMyPower commented:
A simple question, if I'm going to change the dedicated thread pool to `ForkJoinedPool` in future, could it be blocked due to the bug that is fixed in recent JDK versions?

zhanghaou commented:
> Many projects should be affected rather than Pulsar itself. 

Hi, @BewareMyPower 
Could you share which Apache project use the same ForkJoin common pool in various places?

zhanghaou commented:
@lhotari
 
> * An ordinary executor behaves in a different way than the ForkJoinPool. ForkJoinPool can expand the number of threads when it detects blocking and handle this efficiently. This might be relevant in some of the current usage of executors. Switching to use an ordinary executor could cause deadlocks in certain cases where the existing code doesn't deadlock because of ForkJoinPool.

The purpose of this PR is to achieve thread-pool isolation, rather than to compare whether the ForkJoin pool is better than a regular thread pool. For example, in your previous PR https://github.com/apache/pulsar/pull/22589 , you also replaced the ForkJoin common pool with Pulsar’s executor.

> * `MoreExecutor.directExecutor` shouldn't be used at all when `refreshAfterWrite` is used with Caffeine, since it would defeat the purpose of using background refreshing of the cached items. I don't see a reason why someone would want to make an async Caffeine cache synchronous. That doesn't make sense.

https://github.com/apache/pulsar/pull/24745#issuecomment-3298091512  Maybe you miss the explanation here，I had added a new executor.
